### PR TITLE
Unreal Engine Horde Module

### LINKED ISSUE #11: Feature request: Unreal Engine Horde
### Use case

Deployment of [Unreal Engine's Horde CI/CD system](https://github.com/EpicGames/UnrealEngine/blob/5.4/Engine/Source/Programs/Horde/Docs/Deployment.md) following Epic's recommended AWS deployment architecture.

This will enable game developers using Unreal Engine to quickly deploy a purpose built CI/CD pipeline allowing them to take advantage of features such as [Unreal Build Accelerator](https://github.com/EpicGames/UnrealEngine/tree/5.4/Engine/Source/Programs/UnrealBuildAccelerator) (UBA).

### Solution/User Experience

Unreal Horde dashboard, server and agent deployments on ECS, DocumentDB and Elasticache with agents running in an ASG.

- Unreal Engine Horde container needs to securely pulled from Epic Games GHCR using end-user credentials. This ensures that Epic's EULA has been accepted prior to using Horde. See [private-auth ECR documentation.](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/private-auth.html)
- ALB for SSL termination
- Horde dashboard containers
- Horde server containers in both "server" and "worker" mode
- DocumentDB deployment
- Elasticache deployment
- Fully configurable: existing DocDB and Elasticache clusters should be supported, or the module can provision these resources for itself. Same goes for the ECS cluster where these containers are provisioned.

### Alternative solutions

_No response_

### PR DESCRIPTION
**Pull request by henrykie**
**Issue number:**
Closes #11
Addresses #326

## Summary

First version of the Unreal Engine Horde module.

This module provisions the following resources:
- Unreal Engine Horde service on Elastic Container Service powered by AWS Fargate
- Amazon Elasticache with Redis OSS cluster for Horde server
- Amazon DocumentDB cluster for Horde server
- Configurable autoscaling groups to be registered as Horde agents
- SSM Command Document and Ansible playbook for configuring Ubuntu + AL2023 machines as Horde agents
- Security groups, IAM roles, various supporting resources

### Changes

Net new module.

> Please provide a summary of what's being changed

This version provides an easy deployment path for the Unreal Engine Horde CI/CD system on AWS. It can be used to configure the Horde server and N autoscaling groups of agents.

### User experience

> Please share what the user experience looks like before and after this change

Horde server deployment now available.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented

<details>
<summary>Is this a breaking change?</summary>

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created might not be successful.

### CODE REVIEW

File: modules/unreal/horde/variables.tf [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+  description = "The current environment (e.g. Development, Staging, Production, etc.). This will tag ressources and set ASPNETCORE_ENVIRONMENT variable."
+  default     = "Development"
+}
+
+variable "tags" {
+  type = map(any)
+  default = {
+    "IAC_MANAGEMENT" = "CGD-Toolkit"
+    "IAC_MODULE"     = "unreal-horde"
+    "IAC_PROVIDER"   = "Terraform"
```
Comment (jorisdon) [inline modules/unreal/horde/variables.tf:33]: use all-lowercase tag names

File: modules/unreal/horde/variables.tf [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+  }
+  description = "Tags to apply to resources."
+}
+
+variable "debug" {
+  type        = bool
+  description = "Set this flag to enable ECS execute permissions on the Unreal Horde container and force new service deployments on Terraform apply."
+  default     = false
+}
```
Comment (jorisdon) [inline modules/unreal/horde/variables.tf:42]: This is neat and something we should probably do elsewhere as well! :thumbsup:

File: modules/unreal/horde/variables.tf [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+
+variable "docdb_backup_retention_period" {
+  type        = number
+  description = "Number of days to retain backups for DocumentDB cluster."
+  default     = 7
+}
+
+variable "docdb_preferred_backup_window" {
+  type        = string
+  description = "The preferred window for DocumentDB backups to be created."
```
Comment (jorisdon) [inline modules/unreal/horde/variables.tf:332]: Worth noting perhaps that this is in UTC

File: modules/unreal/horde/docdb.tf [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+resource "aws_docdb_cluster_instance" "horde" {
+  count = var.database_connection_string == null ? var.docdb_instance_count : 0
+
+  identifier         = "${var.name}-docdb-${count.index}"
+  cluster_identifier = aws_docdb_cluster.horde[0].id
+  instance_class     = var.docdb_instance_class
+}
+
+resource "aws_docdb_cluster" "horde" {
```
Comment (jorisdon) [inline modules/unreal/horde/docdb.tf:34]: Might be a good idea to also include (variables for) [backup_retention_period](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/docdb_cluster#backup_retention_period) and [preferred_maintenance_window](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/docdb_cluster#preferred_maintenance_window)

**Review by jorisdon — CHANGES_REQUESTED** [review]
Looks great already! Left a few small comments on things to improve.

**Review by kylesomers — APPROVED** [review]

**Review by kylesomers — APPROVED** [review]

**Pull request merged**